src/core/settings.c

```diff
@@ -107,86 +107,204 @@ QuicSettingsCopy(
         Settings->InitialRttMs = ParentSettings->InitialRttMs;
     }
-    if (!Settings->AppSet.MaxAckDelayMs) {
+    if (!Settings->IsSet.MaxAckDelayMs) {
         Settings->MaxAckDelayMs = ParentSettings->MaxAckDelayMs;
     }
-    if (!Settings->AppSet.DisconnectTimeoutMs) {
+    if (!Settings->IsSet.DisconnectTimeoutMs) {
         Settings->DisconnectTimeoutMs = ParentSettings->DisconnectTimeoutMs;
     }
-    if (!Settings->AppSet.KeepAliveIntervalMs) {
+    if (!Settings->IsSet.KeepAliveIntervalMs) {
         Settings->KeepAliveIntervalMs = ParentSettings->KeepAliveIntervalMs;
     }
-    if (!Settings->AppSet.IdleTimeoutMs) {
+    if (!Settings->IsSet.IdleTimeoutMs) {
         Settings->IdleTimeoutMs = ParentSettings->IdleTimeoutMs;
     }
-    if (!Settings->AppSet.HandshakeIdleTimeoutMs) {
+    if (!Settings->IsSet.HandshakeIdleTimeoutMs) {
         Settings->HandshakeIdleTimeoutMs = ParentSettings->HandshakeIdleTimeoutMs;
     }
-    if (!Settings->AppSet.BidiStreamCount) {
+    if (!Settings->IsSet.BidiStreamCount) {
         Settings->BidiStreamCount = ParentSettings->BidiStreamCount;
     }
-    if (!Settings->AppSet.UnidiStreamCount) {
+    if (!Settings->IsSet.UnidiStreamCount) {
         Settings->UnidiStreamCount = ParentSettings->UnidiStreamCount;
     }
-    if (!Settings->AppSet.TlsClientMaxSendBuffer) {
+    if (!Settings->IsSet.TlsClientMaxSendBuffer) {
         Settings->TlsClientMaxSendBuffer = ParentSettings->TlsClientMaxSendBuffer;
     }
-    if (!Settings->AppSet.TlsClientMaxSendBuffer) {
+    if (!Settings->IsSet.TlsClientMaxSendBuffer) {
         Settings->TlsClientMaxSendBuffer = ParentSettings->TlsClientMaxSendBuffer;
     }
-    if (!Settings->AppSet.StreamRecvWindowDefault) {
+    if (!Settings->IsSet.StreamRecvWindowDefault) {
         Settings->StreamRecvWindowDefault = ParentSettings->StreamRecvWindowDefault;
     }
-    if (!Settings->AppSet.StreamRecvBufferDefault) {
+    if (!Settings->IsSet.StreamRecvBufferDefault) {
         Settings->StreamRecvBufferDefault = ParentSettings->StreamRecvBufferDefault;
     }
-    if (!Settings->AppSet.ConnFlowControlWindow) {
+    if (!Settings->IsSet.ConnFlowControlWindow) {
         Settings->ConnFlowControlWindow = ParentSettings->ConnFlowControlWindow;
     }
-    if (!Settings->AppSet.MaxBytesPerKey) {
+    if (!Settings->IsSet.MaxBytesPerKey) {
         Settings->MaxBytesPerKey = ParentSettings->MaxBytesPerKey;
     }
-    if (!Settings->AppSet.ServerResumptionLevel) {
+    if (!Settings->IsSet.ServerResumptionLevel) {
         Settings->ServerResumptionLevel = ParentSettings->ServerResumptionLevel;
     }
 }
 
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+QuicSettingApply(
+    _Inout_ QUIC_SETTINGS* Settings,
+    _In_range_(FIELD_OFFSET(QUIC_SETTINGS, MaxBytesPerKey), >=)
+        uint32_t NewSessionSize,
+    _In_reads_bytes_(NewSessionSize)
+        const QUIC_SETTINGS* NewSettings
+    )
+{
+    // TODO - Input validation
```

Comment:
We need to validate parameters (that need it). Should we fail the call or silently ignore invalid parameters?

Reply:
I think fail the call. Tell consumers up front when they have bugs.

Reply:
Why don't we put a status variable in the settings struct, and if it fails we put the error code in there for the ones that fail. That way a user can know which ones fail. And then return a failed status code. But I'd continue looping and set the rest.

Reply:
I don't understand you. Why would we put a status variable in the struct? We could return an "invalid parameter" error if we want from the SessionOpen call.

Reply:
So you could know which one failed, rather then just one failed. Hardest part about multi setting apps is knowing what failed. Debugging that sucks, especially for an end user.

Reply:
How is putting a single Status variable in the struct any different from returning a status?

Reply:
Because you could loop through the array afterwards and see which one fails. Rather then just getting a generic failure, and having no way to debug other then to just start setting less and less settings until you know which one actually failed.

Reply:
Nevermind. I thought this was passing an array. Forgot this was a bit field and a giant struct.